[css-anchor-position-1] Add parsing support for `anchor()`

RWDavid opened the pull request:
#### 258ed7a6f456590e7c821306211ebd0a41cd6fe8
<pre>
[css-anchor-position-1] Add parsing support for `anchor()`
<a href="https://bugs.webkit.org/show_bug.cgi?id=275063">https://bugs.webkit.org/show_bug.cgi?id=275063</a>
<a href="https://rdar.apple.com/129168625">rdar://129168625</a>

Reviewed by Antti Koivisto.

This patch adds initial parsing support for `anchor()` and introduces
a new CSSValue subclass named CSSAnchorValue. This class is also
implemented as a CSSPrimitiveValue. Note that the parsers for inset
properties (left, top, inset-block-end, etc.) have been modified to
accept `anchor()` as a value, whereas `anchor()` is not accepted as a
&lt;length&gt; in other properties by default.

This patch does not support parsing `anchor()` calls in other CSS math
functions (such as `calc()` and `min()`).

Previously WebKit ignored inset property declarations using `anchor()`.
Now, because these values are properly parsed, any inset properties
using `anchor()` are assigned a default fixed length of 0 during
style resolution. (This causes many spurious test rebaselines.)

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/the-anchor-attribute-003.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-anchor-display-none.tentative-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSAnchorValue.cpp: Added.
(WebCore::CSSAnchorValue::create):
(WebCore::CSSAnchorValue::collectComputedStyleDependencies const):
(WebCore::CSSAnchorValue::customCSSText const):
(WebCore::CSSAnchorValue::equals const):
* Source/WebCore/css/CSSAnchorValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::isValidCSSUnitTypeForDoubleConversion):
(WebCore::isStringType):
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::~CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::create):
(WebCore::CSSPrimitiveValue::unitTypeString):
(WebCore::CSSPrimitiveValue::serializeInternal const):
(WebCore::CSSPrimitiveValue::equals const):
(WebCore::CSSPrimitiveValue::addDerivedHash const):
(WebCore::CSSPrimitiveValue::collectComputedStyleDependencies const):
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSUnits.cpp:
(WebCore::unitCategory):
(WebCore::operator&lt;&lt;):
* Source/WebCore/css/CSSUnits.h:
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isAnchorValue const):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/calc/CSSCalcCategoryMapping.cpp:
(WebCore::hasDoubleValue):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Length.cpp:
(WebCore::CSSPropertyParserHelpers::consumeLengthOrPercent):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Length.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Primitives.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeDashedIdent):
(WebCore::CSSPropertyParserHelpers::consumeAutoOrLengthOrPercent):
(WebCore::CSSPropertyParserHelpers::consumeSide):
(WebCore::CSSPropertyParserHelpers::consumeInsetLogicalStartEnd):
(WebCore::CSSPropertyParserHelpers::consumeAnchor):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertLength):

Canonical link: <a href="https://commits.webkit.org/279891@main">https://commits.webkit.org/279891@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee9c97a809f655d3cb6907ad445a71f5f6c03dc7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54822 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34258 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7401 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58100 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5553 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57122 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41809 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5570 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44403 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3761 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56917 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32387 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47496 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25527 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29177 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4831 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3694 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5051 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59690 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30072 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5200 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51826 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31211 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47576 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51241 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12049 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32222 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30999 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/CSSAnchorValue.cpp:
```diff
@@ -0,0 +1,71 @@
+    else
+        remainingString = m_anchorSide->cssText();
+
+    return makeString("anchor("_s, elementString, remainingString, ")"_s);
```
Using StringBuilder would read better and likely be more efficient too.

nt1m replied:
I've been previously advised to privilege `makeString` over `StringBuilder` for performance when possible by @darinadler and I think @cdumez too.

It is true that `StringBuilder` would read better here though

darinadler replied:
A single call to `makeString` is better for performance than a single use of `StringBuilder`. But using multiple calls to `makeString` is likely worse than just calling `StringBuilder`. It’s a relatively simple refactor to turn this into one large call to `makeString`, which is what I would do. Also please prefer `' '` over `" "_s` and `')'` over `")"_s`.

Here’s a cut that’s probably almost right:
```
auto element = m_anchorElement ? m_anchorElement->cssText() : String { };
auto side = m_anchorSide->cssText();
auto fallback = m_fallback ? m_fallback->cssText() : String { };
if (element.isEmpty() && fallback.isEmpty())
    return makeString("anchor("_s, side, ')');
if (element.isEmpty())
    return makeString("anchor("_s, side, ", "_s, fallback, ')');
if (fallback.isEmpty())
    return makeString("anchor("_s, element, ' ', side, ')');
return makeString("anchor("_s, element, ' ', side, ", "_s, fallback, ')');
```

Here’s a cut at a single `makeString` version that’s probably almost right:
```
auto element = m_anchorElement ? m_anchorElement->cssText() : String { };
auto side = m_anchorSide->cssText();
auto fallback = m_fallback ? m_fallback->cssText() : String { };
auto optionalSpace = element.isEmpty() ? ""_s : " "_s;
auto optionalComma = fallback.isEmpty() ? ""_s : ", "_s;
return makeString("anchor("_s, element, optionalSpace, side, optionalComma, fallback, ')');
```

Can also use `StringBuilder` as @anttijk suggests.

darinadler commented on Source/WebCore/css/CSSAnchorValue.h:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    RefPtr<CSSPrimitiveValue> m_anchorElement;
+    RefPtr<CSSValue> m_anchorSide;
```
Here `m_anchorSide` should be `Ref` instead of `RefPtr` since it can’t be null.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -5812,5 +5821,44 @@ RefPtr<CSSValue> consumeViewTimelineInset(CSSParserTokenRange& range, const CSSP
+    if (!args.size()) {
+        range = rangeCopy;
+        auto anchor = CSSAnchorValue::create(WTFMove(anchorElement), WTFMove(anchorSide), nullptr);
+        return CSSPrimitiveValue::create(anchor);
```
I don’t get why we wrap `CSSAnchorValue` inside a `CSSPrimitiveValue`.

anttijk replied:
I suggested just following `calc()` pattern (which works like this) as deviating from it may result in a large refactoring. There are assumptions that only CSSPrimitiveValues map to Length. But maybe those would be easy enough to untangle?

anttijk replied:
That sort of cleanup should also be done for `calc()` and might be better done separately for both.

RWDavid committed:
[css-anchor-position-1] Add parsing support for `anchor()`
https://bugs.webkit.org/show_bug.cgi?id=275063
rdar://129168625

Reviewed by Antti Koivisto.

This patch adds initial parsing support for `anchor()` and introduces
a new CSSValue subclass named CSSAnchorValue. This class is also
implemented as a CSSPrimitiveValue. Note that the parsers for inset
properties (left, top, inset-block-end, etc.) have been modified to
accept `anchor()` as a value, whereas `anchor()` is not accepted as a
<length> in other properties by default.

This patch does not support parsing `anchor()` calls in other CSS math
functions (such as `calc()` and `min()`).

Previously WebKit ignored inset property declarations using `anchor()`.
Now, because these values are properly parsed, any inset properties
using `anchor()` are assigned a default fixed length of 0 during
style resolution. (This causes many spurious test rebaselines.)

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/the-anchor-attribute-003.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-anchor-display-none.tentative-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSAnchorValue.cpp: Added.
(WebCore::CSSAnchorValue::create):
(WebCore::CSSAnchorValue::collectComputedStyleDependencies const):
(WebCore::CSSAnchorValue::customCSSText const):
(WebCore::CSSAnchorValue::equals const):
* Source/WebCore/css/CSSAnchorValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::isValidCSSUnitTypeForDoubleConversion):
(WebCore::isStringType):
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::~CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::create):
(WebCore::CSSPrimitiveValue::unitTypeString):
(WebCore::CSSPrimitiveValue::serializeInternal const):
(WebCore::CSSPrimitiveValue::equals const):
(WebCore::CSSPrimitiveValue::addDerivedHash const):
(WebCore::CSSPrimitiveValue::collectComputedStyleDependencies const):
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSUnits.cpp:
(WebCore::unitCategory):
(WebCore::operator<<):
* Source/WebCore/css/CSSUnits.h:
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isAnchorValue const):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/calc/CSSCalcCategoryMapping.cpp:
(WebCore::hasDoubleValue):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Length.cpp:
(WebCore::CSSPropertyParserHelpers::consumeLengthOrPercent):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Length.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Primitives.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeDashedIdent):
(WebCore::CSSPropertyParserHelpers::consumeAutoOrLengthOrPercent):
(WebCore::CSSPropertyParserHelpers::consumeSide):
(WebCore::CSSPropertyParserHelpers::consumeInsetLogicalStartEnd):
(WebCore::CSSPropertyParserHelpers::consumeAnchor):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertLength):

Canonical link: https://commits.webkit.org/279891@main
